tss-esapi/src/context/tpm_commands/attestation_commands.rs

```diff
@@ -322,8 +321,120 @@ impl Context {
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use tss_esapi::{Context, TctiNameConf};
+    /// # use std::convert::TryFrom;
+    /// # use tss_esapi::{
+    /// #     abstraction::cipher::Cipher,
+    /// #     interface_types::{
+    /// #         algorithm::{HashingAlgorithm, RsaSchemeAlgorithm},
+    /// #         key_bits::RsaKeyBits,
+    /// #         reserved_handles::Hierarchy,
+    /// #     },
+    /// #     structures::{
+    /// #         RsaExponent, RsaScheme,
+    /// #     },
+    /// #     utils::{create_unrestricted_signing_rsa_public, create_restricted_decryption_rsa_public},
+    /// # };
+    /// use std::convert::TryInto;
+    /// use tss_esapi::{
+    ///     structures::{Data, SignatureScheme},
+    ///     interface_types::session_handles::AuthSession,
+    /// };
+    /// # let mut context =
+    /// #     Context::new(
+    /// #         TctiNameConf::from_environment_variable().expect("Failed to get TCTI"),
+    /// #     ).expect("Failed to create Context");
+    /// let qualifying_data = vec![0xff; 16];
+    /// # let signing_key_pub = create_unrestricted_signing_rsa_public(
+    /// #         RsaScheme::create(RsaSchemeAlgorithm::RsaSsa, Some(HashingAlgorithm::Sha256))
+    /// #         .expect("Failed to create RSA scheme"),
+    /// #     RsaKeyBits::Rsa2048,
+    /// #     RsaExponent::default(),
+    /// # )
+    /// # .expect("Failed to create an unrestricted signing rsa public structure");
+    /// # let sign_key_handle = context
+    /// #     .execute_with_nullauth_session(|ctx| {
+    /// #         ctx.create_primary(Hierarchy::Owner, signing_key_pub, None, None, None, None)
+    /// #     })
+    /// #     .unwrap()
+    /// #     .key_handle;
+    /// let (attest, signature) = context
+    ///     .execute_with_sessions(
+    ///         (
+    ///             Some(AuthSession::Password),
+    ///             Some(AuthSession::Password),
+    ///             None,
+    ///         ),
+    ///         |ctx| {
+    ///             ctx.get_time(
+    ///                 sign_key_handle,
+    ///                 Data::try_from(qualifying_data).unwrap(),
+    ///                 SignatureScheme::Null,
+    ///             )
+    ///         },
+    ///     )
+    ///     .expect("Failed to get tpm time");
+    /// ```
+    pub fn get_time(
+        &mut self,
+        signing_key_handle: KeyHandle,
+        qualifying_data: Data,
+        signing_scheme: SignatureScheme,
+    ) -> Result<(Attest, Signature)> {
+        let mut timeinfo_ptr = null_mut();
+        let mut signature_ptr = null_mut();
+        ReturnCode::ensure_success(
+            unsafe {
+                Esys_GetTime(
+                    self.mut_context(),
+                    ObjectHandle::Endorsement.into(),
+                    signing_key_handle.into(),
+                    self.optional_session_1(),
+                    self.optional_session_2(),
```

Comment:
I think these two are actually required, the first one for the endorsement, the second for the signing key. Looking at `quote` above I wonder if that's a bug, AFAICT the spec asks for one authorization session at least there.

Reply:
*TCG TSS 2.0 Enhanced System API (ESAPI) Specification, Version 1.00, Revision 14, October 1, 2021*
> 11.3.45 Esys_GetTime Functions
> TSS2_DLL_EXPORT TSS2_RC Esys_GetTime_Async(
>    ESYS_CONTEXT *esysContext,
>    ESYS_TR privacyAdminHandle,
>    ESYS_TR signHandle,
>    ESYS_TR privacyAdminHandleSession1,
>    ESYS_TR signHandleSession2,
>    ESYS_TR optionalSession3,
>    TPM2B_DATA const *qualifyingData,
>    TPMT_SIG_SCHEME const *inScheme);

I think you are right only the third session is optional.

Reply:
Regarding `Esys_Quote`.

*TCG TSS 2.0 Enhanced System API (ESAPI) Specification, Version 1.00, Revision 14, October 1, 2021*
> 11.3.42     Esys_Quote Functions
> TSS2_DLL_EXPORT TSS2_RC Esys_Quote_Async(
>     ESYS_CONTEXT *esysContext,
>     ESYS_TR signHandle,
>     ESYS_TR signHandleSession1,
>     ESYS_TR optionalSession2,
>     ESYS_TR optionalSession3,
>     TPM2B_DATA const *qualifyingData,
>     TPMT_SIG_SCHEME const *inScheme,
>    TPML_PCR_SELECTION const *PCRselect);

So it seems Quote is wrong as well because the first session should be required.